configbp/config_test.go

```diff
@@ -0,0 +1,68 @@
+package configbp_test // to break dep cycle
```

Comment:
🔕 even without the potential of dep cycle we should still use external test whenever possible (e.g. we don't need to access any unexported things in the tests) :)

Reply:
We went around and around on this at Google and in the end the consensus is that external tests should only be used when required.

1. There is no consensus that external API tests are actually better, particularly for regression testing. The simplest test that verifies behavior is usually the right one, and sometimes that means using an unexported API.
1. You can still test the exported API only even if you're not in an external test, but the reverse is not true.
1. When using Bazel, if someone decides to make an internal test, it changes what gazelle will generate -- the internal test now takes the "foo_test" target name and shoves the external test to "foo_x_test" which can be confusing, especially if you are tracking things related to the test target over time.

We didn't prevent people from writing external tests, or change them when we found them, but it was not encouraged.

Reply:
I believe 3 is no longer true with new version of gazelle.

I'm not suggesting to use external-only tests. I'm suggesting to use external tests as much as possible, and only use internal tests when you need to test something that's not possible in external tests (so mix them if needed). e.g. use external tests to test public api, plus internal tests to test internal implementations that's not exposed.

it's likely less an issue at google, but I do see people not exposing the api correctly, and not realize that with internal test (because in internal test you still have access to everything). those problems can be avoided with external tests because external tests put you into the api users' shoes.

Reply:
Unfortunately even external tests still have access to `internal/` packages that a user won't have access to.  There is no silver bullet here, so the simplest approach (package-internal tests that can choose to test the external API if they wish) wins out.

Reply:
Especially for unit tests it would prevent you from working with the internal state of a type which you often need to do to test the desired behavior, where as integration tests could live in a different space to begin with.